### DISCUSSION TITLE
Why does the push notification payload include an access token?

### DISCUSSION THREAD

**Discussion by ixsiid** [mastodon/mastodon#27084] [Q&A]
What is the client expected to use the access token for?
I also believe that the security risk lies in the fact that the server sends the access token out of the OAuth flow.

**Comment by ryanmio** [comment]
The token is often used to authorize the client app to perform actions like fetching additional data

**Reply by ixsiid** [reply]
The client app already holds the token.
(since it is needed to perform push/subscription)
Does that mean that some service workers cannot hold the token

**Reply by ryanmio** [reply]
Sorry I misunderstood your question. You're right, the client app already holds the token, making it redundant to also send it via push notifications. 

Can you confirm where you observed the access token in the push notification payload? Are you sure it's not encrypted or secured in any way? Also, which service/system are you referring to?

**Reply by ixsiid** [reply]
I do not understand all the workings of mastodon.

Here, it appears that access tokens are used to create objects for push notifications.
https://github.com/mastodon/mastodon/blob/88fa8e710adf103dee289b2a9cc691a410f97c21/app/serializers/web/notification_serializer.rb#L11

Here is a small instance, but the access token is actually delivered.
https://friends.cafe/

It will take some time, but I will investigate some larger instances as well.

**Comment by ryanmio** [comment]
> Does that mean that some service workers cannot hold the token

Service workers can hold tokens, but it's not a best practice
